pkg/multishare_reconciler/reconciler_test.go

```diff
@@ -0,0 +1,313 @@
+				IsRegional:  test.isRegional,
+			},
+			cloud: &cloud.Cloud{
+				Zone: test.cloudZone,
+			},
+		}
+		instances, shares, shareMap, _ := recon.managedInstanceAndShare(test.instances, test.shares)
+		if len(instances) != len(test.expectedInstanceNames) {
+			t.Errorf("expecting %d instances, got %d", len(test.expectedInstanceNames), len(instances))
+		}
+		if len(shares) != len(test.expectedShareNames) {
+			t.Errorf("expecting %d shares, got %d", len(test.expectedShareNames), len(shares))
+		}
+		for _, instance := range instances {
+			if !slices.Contains(test.expectedInstanceNames, instance.Name) {
+				t.Errorf("expected instance with names %v but got %q", test.expectedInstanceNames, instance.Name)
+			}
+		}
+		for _, share := range shares {
+			if !slices.Contains(test.expectedShareNames, share.Name) {
+				t.Errorf("expected instances with names %v but got %q", test.expectedShareNames, share.Name)
+			}
+		}
+		for instanceURI, shareList := range shareMap {
+			for _, share := range shareList {
+				if !slices.Contains(test.expectedinstanceShareMapping[instanceURI], share.Name) {
+					t.Errorf("instance %q should map to %v but got %q", instanceURI, test.expectedinstanceShareMapping[instanceURI], share.Name)
+				}
+			}
+		}
+	}
+}
+
+func TestInstanceEmpty(t *testing.T) {
+	cases := []struct {
+		name     string
+		instance *v1alpha1.InstanceInfo
+		expected bool
+	}{
+		{
+			name: "empty instance",
+			instance: &v1alpha1.InstanceInfo{
+				Status: &v1alpha1.InstanceInfoStatus{
+					ShareNames: []string{},
+				},
+			},
+			expected: true,
+		},
+		{
+			name: "non-empty instance",
+			instance: &v1alpha1.InstanceInfo{
+				Status: &v1alpha1.InstanceInfoStatus{
+					ShareNames: []string{"share1"},
+				},
+			},
+			expected: false,
+		},
+		{
+			name:     "nil status instance",
+			instance: &v1alpha1.InstanceInfo{},
+			expected: false,
+		},
+	}
+
+	for _, test := range cases {
+		empty := instanceEmpty(test.instance)
+		if empty != test.expected {
+			t.Errorf("case %s expects %v but got %v", test.name, test.expected, empty)
+		}
+	}
+}
+
+func TestMaybeAddCleanupFinalizer(t *testing.T) {
+	now := metav1.Now()
+	cases := []struct {
+		name     string
+		instance *v1alpha1.InstanceInfo
+		expected bool
+	}{
+		{
+			name:     "no deletionTimestamp",
+			instance: &v1alpha1.InstanceInfo{},
+			expected: true,
+		},
+		{
+			name: "deletionTimestamp set",
+			instance: &v1alpha1.InstanceInfo{
+				ObjectMeta: metav1.ObjectMeta{
+					DeletionTimestamp: &now,
+				},
+			},
+			expected: false,
+		},
+		{
+			name: "replace finalizer",
+			instance: &v1alpha1.InstanceInfo{
+				ObjectMeta: metav1.ObjectMeta{
+					Finalizers: []string{"thisNThat"},
+				},
+			},
+			expected: true,
+		},
+	}
+
+	for _, test := range cases {
+		instanceInfo, updated := maybeAddCleanupFinalizer(test.instance)
+		if updated != test.expected {
+			t.Errorf("case %s expects %v but got %v", test.name, test.expected, updated)
```

Comment:
wouldn't Fatalf also stops the whole test?

Reply:
It will, I thought you wanted to stop the comparison if instanceInfo is not updated as expected, then adding a "continue" should be enough. NVM if my understanding was wrong.